src/ShaderGlossy.h

```diff
@@ -25,23 +25,57 @@ class CShaderGlossy : public CShaderPhong {
 	virtual ~CShaderGlossy(void) = default;
 
 	virtual Vec3f shade(const Ray& ray) const override {
-		Vec3f res = CShaderPhong::shade(ray);
 
-		Vec3f normal = ray.hit->getNormal(ray);									// shading normal
-
-		// --- PUT YOUR CODE HERE ---
-
-		Ray reflected;
-		reflected.org = ray.org + ray.dir * ray.t;
-		reflected.dir = normalize(ray.dir - 2 * normal.dot(ray.dir) * normal);
-
-		Vec3f reflection = m_scene.RayTrace(reflected);
-
-		res = 0.5 * res + 0.5 * reflection;
-
-		return res;
+	    // shading normal
+        int nSamples = m_pSampler->getNumSamples();
+        Vec3f agg = Vec3f::all(0);
+        Vec3f res = CShaderPhong::shade(ray);
+        Vec3f normal = ray.hit->getNormal(ray);
+        int k = 0;
+        for(int i = 0; i < nSamples; i++) {
+            // deep copy of normal.
+            auto normalCpy = Vec3f(normal);
+            Vec2f sample = m_pSampler -> getSample(i);
+            auto res2 = cwsReflection(sample, normalCpy, ray, nSamples);
+            if (res2.has_value())
+            {
+                k++;
+                agg += res2.value() * m_glossiness;
```

Comment:
it is a mis-use go the glossiness effect. This implementation just makes glossiness to vanish, but the m_glossiness parameter should modulate how much the normal is deviated - thus, for m_glossiness close to 1, the samples should be concentrated on top of the hemisphere, whereas for small m_glossiness close to 0, the samples should be equally distributed over the hemisphere